src/handlers/worker_handler.py

```diff
@@ -0,0 +1,198 @@
+    @authenticate
+    @authenticate_worker
+    @schema.validate(
+        output_schema={
+            "type": "object",
+            "properties": {
+                api_key.JOB_ID: {"type": "string"},
+                api_key.STAGES: {
+                    "type": "array",
+                    "items": grading_job_def,
+                },
+                api_key.STUDENTS: {
+                    "type": "array",
+                    "items": {"type": "object"},
+                }
+            },
+            "required": [api_key.JOB_ID, api_key.STAGES],
+            "additionalProperties": False
+        }
+    )
+    def get(self, worker_id):
+        db_resolver = self.get_db()
+        job_queue = self.get_queue()
+
+        try:
+            job = job_queue.get_nowait()
+            job_id = job.get(api_key.JOB_ID)
+
+            db_resolver.get_grading_job_collection().update_one({db_key.ID: ObjectId(job_id)},
+                                                                {"$set": {db_key.STARTED: get_time()}})
+
+            db_resolver.get_worker_node_collection().update_one({db_key.ID: ObjectId(worker_id)},
+                                                                {"$set": {db_key.RUNNING_JOB: job_id}})
+
+            return job
+        except Exception as e:
+            self.set_status(QUEUE_EMPTY_CODE)
+            return {api_key.JOB_ID: 'no id', api_key.STAGES: []}
+
+    # POST used to update grading job status upon completion
+    @authenticate
+    @authenticate_worker
+    @schema.validate(
+        input_schema={
+            "type": "object",
+            "properties": {
+                api_key.JOB_ID: {"type": "string"},
+                api_key.SUCCESS: {"type": "boolean"},
+                api_key.RESULTS: {
+                    "type": "array",
+                    "items": {"type": "object"},
+                },
+                api_key.LOGS: {"type": "object"}
+            },
+            "required": [api_key.JOB_ID, api_key.SUCCESS, api_key.RESULTS, api_key.LOGS],
+            "additionalProperties": False
+        }
+    )
+    def post(self, worker_id):
+        db_handler = self.get_db()
+        worker_nodes_collection = db_handler.get_worker_node_collection()
+        jobs_collection = db_handler.get_grading_job_collection()
+        job_logs_collection = db_handler.get_job_logs_collection()
+
+        # update worker node: remove this job from its currently running jobs
+        worker_nodes_collection.update_one({db_key.ID: ObjectId(worker_id)},
+                                           {"$set": {db_key.RUNNING_JOB: None}})
+
+        # check if the job exists
+        job_id = self.body.get(api_key.JOB_ID)
+        job = self.get_grading_job(job_id)
+        if job is None:
+            return
+
+        # update job: finished_at and result
+        assert db_key.CREATED in job
+        assert db_key.QUEUED in job
+        assert db_key.STARTED in job
+        assert db_key.FINISHED not in job
```

Comment:
Are you trying to do integrity checks with asserts or do you want to fail fast if the keys don't match up?

Reply:
I feel like we should not be using asserts here in the first place. If this is a JSON payload, why aren't we just using a JSON validation and returning 400 if the validation fails?

Reply:
To make things clearer, `job` here is a mongo document. Does not come as a payload so can not use JSON validation. Those keywords are used to decipher the state of the job. When the job is created, it only has the CREATED key, when its queued, QUEUED is added and similarly for STARTED and FINISHED

Reply:
Should we return a 400 or something else instead of an exception? Or does that get handled by tornado automatically (in that case we should document that there is a code difference between running this in "debug" mode versus production mode.)

Reply:
Yes returning a 400+ code would be better. I will fix this

Reply:
Handled all possible cases with appropriate actions. Should be fine now. No more asserts.